connector/ws/src/lib.rs

```diff
@@ -97,24 +122,157 @@ impl<S: connector::Supervisor> WsConnectorInner<S> {
                 .await
                 .map_err(|e| WsConnectorError::Connection(e))?;
 
-        let _ = socket_auth::authenticate_as_supervisor(&mut ws, supervisor_id, signing_key)
+        let _ = socket_auth::authenticate_as_supervisor(&mut ws, self.supervisor_id, signing_key)
             .await
             .map_err(|e| WsConnectorError::Authentication(e))?;
 
-        Ok(Self {
-            supervisor_id,
-            config,
-            supervisor,
-            connection: ws,
-        })
+        Ok(ws)
+    }
+    async fn handle(&self, message: switchboard_supervisor::Message) {
+        match message {
+            switchboard_supervisor::Message::StartJob(start_job_request) => {
+                let job_id = start_job_request.job_id;
+                if let Some(supervisor) = self.supervisor.upgrade() {
+                    // TODO: timeout
+                    if let Err(error) =
+                        connector::Supervisor::start_job(&supervisor, start_job_request).await
+                    {
+                        self.report_job_error(job_id, error).await;
+                        // self.update_tx
+                        //     .send(switchboard_supervisor::Message::SupervisorEvent(
+                        //         SupervisorEvent::ReportJobError { job_id, error },
+                        //     ))
+                        //     .unwrap();
+                    }
+                }
+            }
+            switchboard_supervisor::Message::StopJob(stop_job_request) => {
+                let job_id = stop_job_request.job_id;
+                // TODO: timeout
+                if let Some(supervisor) = self.supervisor.upgrade() {
+                    if let Err(error) =
+                        connector::Supervisor::stop_job(&supervisor, stop_job_request).await
+                    {
+                        self.report_job_error(job_id, error).await;
+                        // self.update_tx
+                        //     .send(switchboard_supervisor::Message::SupervisorEvent(
+                        //         SupervisorEvent::ReportJobError { job_id, error },
+                        //     ))
+                        //     .unwrap();
+                    }
+                }
+            }
+            switchboard_supervisor::Message::StatusRequest(switchboard_supervisor::Request {
+                request_id,
+                message: (),
+            }) => {
+                let status = self.last_updated_status.lock().await.clone();
+                self.update_tx
+                    .send(switchboard_supervisor::Message::StatusResponse(Response {
+                        response_to_request_id: request_id,
+                        message: status,
+                    }))
+                    .unwrap();
```

Comment:
If it would help to simply be able to query the status from the underlying `connector::Supervisor` instance instead, we can also add that API. But I'm not opposed to simply caching it on update and returning it here as well.

Reply:
Originally that was exactly the plan, but wasn't sure how to implement it at the `Supervisor`; additionally, I can't actually think of whether it provides any advantage over e.g. cache-on-change.
If we can request status, it makes sense to be able to request status at any point in the job execution process; however, this implies that the status request runs concurrently with any other ongoing tasks.
In order to have a concurrent status request and job control request, that would require extra synchronization inside the supervisor, and I found it easier to just perform it in the `SupervisorConnector`, since I'm not super familiar with the supervisor codebases (plus it allows easier drop-in support, like we saw today).

I also remember that there were some other issues I ran into, but I *think* I resolved those by making `WsConnector` into a wrapper around an `Arc` of an inner type that took `self: &Arc<Self>` instead of `&self`.